From 91a187df94ca793411afcd5ab51a40635dea9376 Mon Sep 17 00:00:00 2001
From: Ramya Gnanasekar <quic_rgnanase@quicinc.com>
Date: Wed, 2 Mar 2022 19:38:50 +0530
Subject: [PATCH] ath12k: Add support to detect MIC failure

In Beriliyum, crypto hw will set TKIP error and Decrypt error
when it detects MIC failure and ICV failure. In this case, RXDMA
will give higher priority to decrypt error and sets the error code
as decrypt error(3). Ath12k driver expects TKIP MIC error (4) in case
of mic failure and hence no action is taken against the STA which
injected the corrupted packet.

To overcome the scenario, TKIP error bit set in msdu_end is checked
when decrypt or tkip error is reported by RXDMA and necessary action
is taken.

Signed-off-by: Ramya Gnanasekar <quic_rgnanase@quicinc.com>

--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -3750,14 +3750,21 @@ static bool ath12k_dp_rx_h_rxdma_err(str
 				     struct ieee80211_rx_status *status)
 {
 	struct ath12k_skb_rxcb *rxcb = ATH12K_SKB_RXCB(msdu);
+	struct hal_rx_desc *rx_desc = (struct hal_rx_desc*)msdu->data;
 	bool drop = false;
+	u32 err_bitmap;
 
 	ar->ab->soc_stats.rxdma_error[rxcb->err_code]++;
 
 	switch (rxcb->err_code) {
+	case HAL_REO_ENTR_RING_RXDMA_ECODE_DECRYPT_ERR:
 	case HAL_REO_ENTR_RING_RXDMA_ECODE_TKIP_MIC_ERR:
-		ath12k_dp_rx_h_tkip_mic_err(ar, msdu, status);
-		break;
+		err_bitmap = ath12k_dp_rx_h_mpdu_err(rx_desc);
+		if (err_bitmap & DP_RX_MPDU_ERR_TKIP_MIC) {
+			ath12k_dp_rx_h_tkip_mic_err(ar, msdu, status);
+			break;
+		}
+		fallthrough;
 	default:
 		/* TODO: Review other rxdma error code to check if anything is
 		 * worth reporting to mac80211
